cmd/snap-confine/snap-confine.c

```diff
@@ -347,11 +333,104 @@ int main(int argc, char **argv) {
     debug("ruid: %d, euid: %d, suid: %d", real_uid, effective_uid, saved_uid);
     debug("rgid: %d, egid: %d, sgid: %d", real_gid, effective_gid, saved_gid);
 
-    // snap-confine needs to run as root for cgroup/udev/mount/apparmor/etc setup.
-    if (effective_uid != 0) {
-        die("need to run as root or suid");
+    struct sc_apparmor apparmor;
+    sc_init_apparmor_support(&apparmor);
+    if (!apparmor.is_confined && apparmor.mode != SC_AA_NOT_APPLICABLE && real_uid != 0) {
+        // Refuse to run when this process is running unconfined on a system
+        // that supports AppArmor when the effective uid is root and the real
+        // id is non-root.  This protects against, for example, unprivileged
+        // users trying to leverage the snap-confine in the core snap to
+        // escalate privileges.
+        errno = 0;  // errno is insignificant here
+        die("snap-confine has elevated permissions and is not confined"
+            " but should be. Refusing to continue to avoid"
+            " permission escalation attacks\n"
+            "Please make sure that the snapd.apparmor service is enabled and started.");
+    }
+
+    sc_debug_capabilities("initial caps");
+
+    static const cap_value_t snap_confine_caps[] = {
+        CAP_DAC_OVERRIDE,     // poking around as a regular user
+        CAP_DAC_READ_SEARCH,  // same as above
+        CAP_SYS_ADMIN,        // mounts, unshare
+        CAP_SYS_CHROOT,       // pivot_root into a new root
+        CAP_CHOWN,            // file ownership
+        CAP_FOWNER,           // to create tmp dir with sticky bit
+        CAP_SYS_PTRACE,       // to inspect the mount namespace of PID1
+        CAP_SETUID,           // assume user identity
+        CAP_SETGID,           // and group identity, e.g. switching to root when running s-u-n
```

Comment:
are these still relevant? at least the comments need some tweaking

Reply:
since we no longer switch uid/gid those permissions are no longer needed